### PR TITLE
♻️ Refactor lambda code to deploy Container lambda

### PR DESCRIPTION
**Pull request by devbyaccident**
Changes made in this commit:
1. Added GitHub actions workflows for [Bridgecrew](https://github.com/bridgecrewio/bridgecrew-action), [Yor](https://github.com/bridgecrewio/yor), [Terraform formatting](https://github.com/hashicorp/setup-terraform) and the [Github super-linter](https://github.com/github/super-linter).
2. Updated gitignore file to match python recommended [gitignore](https://github.com/github/gitignore/blob/main/Python.gitignore) with some additions for terraform
3. Add Dockerfile to use [AWS Lambda base image](https://docs.aws.amazon.com/lambda/latest/dg/python-image.html#python-image-create). The lambda now deploys as a [container image](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-container-image-support/) instead of a layered function. 
4. Adding terraform configuration for lambda and ancillary resources
5. Change Python code to run under AWS lambda handler
6. Adding standard build and deploy scripts

Further changes will need to be made to to the Jenkins pipeline to use new deployment scripts.

### CODE REVIEW

File: deployment/terraform/lambda.tf [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  name = var.ecr_name
+}
+
+resource "aws_lambda_function" "kf_lambda_function" {
+  function_name = "${var.application}-${var.environment}"
+  description   = "This lambda receive events from the SQS queue that contains users from persona."
+  role          = aws_iam_role.aws_lambda_role.arn
+
+  package_type = "Image"
+  image_uri    = "${data.aws_ecr_repository.ecr.repository_url}:${var.lambda_version}"
```
Comment (devbyaccident) [inline deployment/terraform/lambda.tf:22]: The `lambda_version` variable is set in the build and infra scripts, and is pulled from the git commit hash.

File: deployment/terraform/sqs.tf [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# sqs.tf
+
+resource "aws_sqs_queue" "kf_lambda_sqs" {
+  name                       = "lambda-${var.application}-${var.environment}-sqs"
+  delay_seconds              = 0
+  max_message_size           = 262144
+  message_retention_seconds  = 86400
+  receive_wait_time_seconds  = 10
+  visibility_timeout_seconds = 60
```
Comment (devbyaccident) [inline deployment/terraform/sqs.tf:11]: These were hard-coded in the original configuration, so I kept it the same here, but this can also be parameterized.

File: src/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+boto3==1.9.219
+elasticsearch7==7.13.0
```
Comment (devbyaccident) [inline src/requirements.txt:2]: These are the requirements for the application. There is a separate file for requirements for testing.

File: requirements.txt [hunk 4]
```diff
@@ -1,2 +1,3 @@
-boto3==1.9.219
-elasticsearch7==7.13.0
\ No newline at end of file
+pytest==5.1.1
+black==22.10.0
+pylint==2.15.5
```
Comment (devbyaccident) [inline requirements.txt:3]: These are the requirements for testing the application. There is a separate file for requirements for production.

File: deployment/terraform/iam.tf [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    actions = [
+      "s3:GetObject",
+      "s3:ListBucket",
+      "s3:GetBucketLocation",
+    ]
+
+    resources = [
+      "arn:aws:s3:::${var.payload_bucket}/*",
+      "arn:aws:s3:::${var.payload_bucket}",
```
Comment (devbyaccident) [inline deployment/terraform/iam.tf]: I wasn't sure if this should be left in or not for the function itself, or if this was only used for uploading the lambda zip in the previous configuration. I left it in out of caution, but we might be able to remove this and slim down this profile too.

File: deployment/terraform/sqs.tf (hunk 2 above)
Comment (alubneuski) [reply deployment/terraform/sqs.tf:11]: It is probably fine.

File: deployment/terraform/variables.tf [hunk 6]
```diff
@@ -0,0 +1,57 @@
+variable "environment" {
+  description = "Dev/QA/QC/Prd"
```
Comment (alubneuski) [inline deployment/terraform/variables.tf]: We should probably change it to : "dev,qa,prd,service"

File: deployment/terraform/subnets.tf [hunk 7]
```diff
@@ -0,0 +1,28 @@
+data "aws_subnet" "subnet1" {
```
Comment (alubneuski) [inline deployment/terraform/subnets.tf]: You can also use networking module:
```
module "network" {
  source        = "git@github.com:kids-first/aws-infra-network-module"
  organization  = var.organization
  environment   = var.environment
  subnet_prefix = "apps"
  azs           = var.azs
}
```
There are the outputs from the module:
https://github.com/kids-first/aws-infra-network-module/blob/master/outputs.tf

Comment (alubneuski) [reply deployment/terraform/subnets.tf]: ```azs = ["a","b"]```

File: deployment/terraform/lambda.tf [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  environment {
+    variables = {
+      es_host   = var.es_host
+      es_scheme = "https"
+      es_port   = "443"
+    }
+  }
+
+  vpc_config {
+    subnet_ids         = [data.aws_subnet.subnet1.id, data.aws_subnet.subnet2.id, data.aws_subnet.subnet3.id, data.aws_subnet.subnet4.id]
```
Comment (alubneuski) [inline deployment/terraform/lambda.tf]: If you use the module, here you can just have:
```subnet_ids = module.network.private_subnets```

File: deployment/terraform/subnets.tf (hunk 7 above)
Comment (devbyaccident) [reply deployment/terraform/subnets.tf]: I thought about that, but I didn't want to reference a private module in a public repo. I can still make the change, I don't think it'd be harmful.

File: deployment/terraform/lambda.tf (hunk 8 above)
Comment (devbyaccident) [reply deployment/terraform/lambda.tf]: Fixed in https://github.com/kids-first/kf-lambda-member-update/pull/15/commits/5e485fa4bb17409225b69a0a8790e9287837140f

File: deployment/terraform/subnets.tf (hunk 7 above)
Comment (devbyaccident) [reply deployment/terraform/subnets.tf]: Fixed in https://github.com/kids-first/kf-lambda-member-update/pull/15/commits/5e485fa4bb17409225b69a0a8790e9287837140f

File: deployment/terraform/variables.tf (hunk 6 above)
Comment (devbyaccident) [reply deployment/terraform/variables.tf]: Fixed in https://github.com/kids-first/kf-lambda-member-update/pull/15/commits/863c16d1a98cda0f86463dabd971ba715dc8162e

File: .github/workflows/linter.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+      - name: Lint Code Base
+        uses: github/super-linter@v4
+        env:
+          VALIDATE_ALL_CODEBASE: false
+          DEFAULT_BRANCH: master
+          VALIDATE_BASH: false
+          VALIDATE_JSCPD: false
+          VALIDATE_MARKDOWN: false
+          VALIDATE_PYTHON_ISORT: false
```
Comment (devbyaccident) [inline .github/workflows/linter.yml:26]: I ended up removing a few of the python linters as we're already using black to format python code. Also removed tflint because it requires init of the network module, which is private.

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+  "python.formatting.provider": "black"
```
Comment (devbyaccident) [inline .vscode/settings.json:2]: Using black to format code is now configured in VSCode for anyone who clones the repo

File: deployment/terraform/cloudwatch.tf [hunk 11]
```diff
@@ -0,0 +1,15 @@
+resource "aws_cloudwatch_log_group" "app" {
+  #checkov:skip=CKV_AWS_158:TODO-Add encryption to cloudwatch log group
```
Comment (devbyaccident) [inline deployment/terraform/cloudwatch.tf:2]: This will need to be a separate PR to try to implement [CKV_AWS_158](https://docs.bridgecrew.io/docs/ensure-that-cloudwatch-log-group-is-encrypted-by-kms), and we need to figure out if we should make a separate key for it for the account or use a separate key per application.

File: deployment/terraform/lambda.tf [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+data "aws_ecr_repository" "ecr" {
+  name = var.ecr_name
+}
+
+resource "aws_lambda_function" "kf_lambda_function" {
+  #checkov:skip=CKV_AWS_272:No definition for this rule
+  #checkov:skip=CKV_AWS_116:TODO-Add dead-letter queue
+  #ts:skip=AWS.LambdaFunction.LM.MEIDUM.0063 Policy is defined in attached lambda role
```
Comment (devbyaccident) [inline deployment/terraform/lambda.tf:13]: I'm skipping a few rules here for checkov and terrascan, but I think the only one we will need to address at some point is [CKV_AWS_116](https://docs.bridgecrew.io/docs/ensure-that-aws-lambda-function-is-configured-for-a-dead-letter-queue-dlq) and that may not be a priority.

File: deployment/terraform/sqs.tf [hunk 13]
```diff
@@ -0,0 +1,51 @@
+# sqs.tf
+
+resource "aws_sqs_queue" "kf_lambda_sqs" {
+  #checkov:skip=CKV_AWS_27:TODO-Will need a seperate story for adding encryption to SQS
+  #ts:skip=AWS.SQS.NetworkSecurity.High.0570 TODO-Will need a seperate story for adding encryption to SQS
```
Comment (devbyaccident) [inline deployment/terraform/sqs.tf:5]: In it's current state, encrypting SQS breaks the lambda. Adding these skips until we can address that in a separate story.

File: deployment/terraform/securitygroups.tf [hunk 14]
```diff
@@ -0,0 +1,29 @@
+data "aws_vpc" "vpc_lookup" {
+  tags = {
+    Name = "${var.organization}-apps-${var.environment}-${data.aws_region.current.name}-vpc"
```
Comment (devbyaccident) [inline deployment/terraform/securitygroups.tf:3]: I kept this as is from the config repo, but we can also pass in a VPC ID instead and lookup from that. No need to expose naming.

File: .github/workflows/bc_check.yml [hunk 15]
```diff
@@ -0,0 +1,22 @@
+---
+name: Run Bridgecrew
+on:
+  pull_request:
+    branches: [master, feature/cbl/deployment-config]
```
Comment (evans-g-crsj) [inline .github/workflows/bc_check.yml]: Should we support that branch `feature/cbl/deployment-config`?

File: .github/workflows/check_pull_request.yml [hunk 16]
```diff
@@ -2,7 +2,7 @@ name: Check Pull Request Quality
 
 on:
   pull_request:
-    branches: [master]
+    branches: [master, feature/cbl/deployment-config]
```
Comment (evans-g-crsj) [inline .github/workflows/check_pull_request.yml]: Same question as above?

File: .github/workflows/linter.yml [hunk 17]
```diff
@@ -0,0 +1,28 @@
+---
+name: Lint Code Base
+on:
+  pull_request:
+    branches: [master, feature/cbl/deployment-config]
```
Comment (evans-g-crsj) [inline .github/workflows/linter.yml]: Same as comment as earlier.

File: .github/workflows/terraform.yml [hunk 18]
```diff
@@ -0,0 +1,23 @@
+name: Terraform Plan
+on:
+  pull_request:
+    branches: [ master, feature/cbl/deployment-config ]
```
Comment (evans-g-crsj) [inline .github/workflows/terraform.yml]: :arrow_up:

File: .github/workflows/yor.yml [hunk 19]
```diff
@@ -0,0 +1,23 @@
+name: IaC trace
+
+on:
+  pull_request:
+    branches: [ master, feature/cbl/deployment-config ]
```
Comment (evans-g-crsj) [inline .github/workflows/yor.yml]: :arrow_up:

File: .github/workflows/bc_check.yml (hunk 15 above)
Comment (devbyaccident) [reply .github/workflows/bc_check.yml]: Since this is a new test that is not currently on the main branch, this needs to be present for running the actions on this branch, I'm going to remove it before merging.

File: .github/workflows/check_pull_request.yml (hunk 16 above)
Comment (devbyaccident) [reply .github/workflows/check_pull_request.yml]: https://github.com/kids-first/kf-lambda-member-update/pull/15#discussion_r1038792232

File: .github/workflows/linter.yml (hunk 17 above)
Comment (devbyaccident) [reply .github/workflows/linter.yml]: https://github.com/kids-first/kf-lambda-member-update/pull/15#discussion_r1038792232

File: .github/workflows/terraform.yml (hunk 18 above)
Comment (devbyaccident) [reply .github/workflows/terraform.yml]: https://github.com/kids-first/kf-lambda-member-update/pull/15#discussion_r1038792232

File: .github/workflows/yor.yml (hunk 19 above)
Comment (devbyaccident) [reply .github/workflows/yor.yml]: https://github.com/kids-first/kf-lambda-member-update/pull/15#discussion_r1038792232

File: deployment/terraform/cloudwatch.tf (hunk 11 above)
Comment (alubneuski) [reply deployment/terraform/cloudwatch.tf:2]: Separate key per application, might be a better option.

File: deployment/terraform/iam.tf [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+    effect = "Allow"
+
+    actions = [
+      "logs:CreateLogGroup",
+      "logs:CreateLogStream",
+      "logs:PutLogEvents",
+    ]
+
+    resources = [
+      "arn:aws:logs:*:*:*",
```
Comment (alubneuski) [inline deployment/terraform/iam.tf]: Do you think, we can narrow it down to an account scale?

**Review by alubneuski — COMMENTED** [review]
README file needs to be updated. I could not follow instructions in the README.md file to setup the environment.

File: deployment/terraform/iam.tf (hunk 20 above)
Comment (devbyaccident) [reply deployment/terraform/iam.tf]: Fixed in https://github.com/kids-first/kf-lambda-member-update/pull/15/commits/38c10495ed2bdd3842dad3848124559196a0844e, narrowed it down to the log group created in the same configuration.

**Comment by devbyaccident** [comment]
> Separate key per application, might be a better option.

@alubneuski So are you thinking a separate key for cloudwatch per application, or one KMS key per app shared between all of it's resources? This app already makes a [KMS key](https://github.com/kids-first/kf-lambda-member-update/blob/38c10495ed2bdd3842dad3848124559196a0844e/deployment/terraform/lambda.tf#L17), but that's being used for lambda function.

**Comment by alubneuski** [comment]
Great job! 
The only note, I would strongly suggest keeping the naming for the scripts according to "Scripts Rule Them All" standard:
https://github.com/github/scripts-to-rule-them-all/tree/master/script
It is not super easy to understand what run-sh-with-pip-requirements.sh means, (is it bootstrap or start server or something else).

**Review by alubneuski — APPROVED** [review]
💯

**Comment by devbyaccident** [comment]
> The only note, I would strongly suggest keeping the naming for the scripts according to "Scripts Rule Them All" standard:
https://github.com/github/scripts-to-rule-them-all/tree/master/script
It is not super easy to understand what run-sh-with-pip-requirements.sh means, (is it bootstrap or start server or something else).

Fixed in https://github.com/kids-first/kf-lambda-member-update/pull/15/commits/87870dd22ed1fb25b5e1d8054d5b70bd4463fc3d. Thanks for the 👍 ! 

I'll get this fixed-up and merged.
